README.md

```diff
@@ -70,6 +70,86 @@ impl IdAllocator {
 }
 ```
 
+## AddressAllocator
+
+### Design address allocator
+
+This allocator should be used to allocate memory slots, either for memmory-mapped
+I/O or for port I/O or any other kind of address space.
+
+The struct that defined the AddressAllocator contains the ends of the interval
+that is managed and an instance of `interval_tree` as an allocation engine. We
+chose to use `interval_tree` for memory slots allocation because it provides
+better creation performance and more query speed.
+
+```rust
+/// Address allocator representation
+pub struct AddressAllocator {
+    // Begining of the address space that we want to manage.
+    base: u64,
+    // End of the address space that we want to manage.
+    end: u64,
+    // Allocation engine, the allocation logic that is interfaced
+    // by AddressAllocator
+    interval_tree: IntervalTree,
+}
+````
+
+The `AddressAllocator` struct implements methods for allocating and releasing
+memory slots from the managed address space.
+
+```rust
+impl AddressAllocator {
+    ///  Creates a new AddressAllocator object with an empty IntervalTree
+    pub fn new(base: u64, size: u64) -> std::result::Result<Self, Error> { }
```

Comment:
The struct has the `end` field, while the constructor has `size`. Is this by design, i.e. does `base` and `size` in the constructor get translated to `base` and `end` in the struct?

Reply:
Also, looking at the `allocate()` method below, the return type should probably be `Result<Self>`.

Reply:
The allocate method return an interval [a, b] that correspond to the constraint, I do not believe it should return Result<Self>

Reply:
I see how my comment was confusing, sorry :( What I meant is that `new()` should return `Result<Self>` similar to how `allocate()` returns `Result<Range>`. So, `Result<Self>` instead of `std::result::Result<Self, Error>`.